README.md
soundgals
=========

Computer Audio Development

This is an update to the XBMC/Kodi plug in; Qobuz XBMC.

The update adds a new "Audiophile" option under the add-on settings within the XBMC/Kodi GUI.

When this option is selected via its radio button, each time a user selects either an album or playlist from within
Qobuz XBMC a file named qobuzNow.m3u8 will be written as a playlist which contains all the urls necessary to
play the files from this playlist. If the file qobuzNow.m3u8 doesn't exist it will be created. The file is written
to the users Music directory within his/her home directory of the OS.

The user must have designated a program to open all files of type .m3u8 within his/her operating system.

This should be a program such as foobar2000, HQPlayerDesktop, J River Media center or VLC, since all of these support
flac streaming.

As long as a default program which supports flac streaming is chosen in the os by the user, the qobuzNow.m3u8 file
will be opened by that program.

Turning the "Audiophile" feature off will cause Qobuz XBMC to revert to its default behaviour of playing Qobuz
streams through XBMC.

Version 1.1.1 Now adds the support for individual tracks to take advantage of this "Audiophile" option.
For this to work you must turn off "Play the next song automatically" under Sytem -> Settings -> Music -> Playback.

Version 1.1.2 Adds choice within Audiophile settings between Gapless and non-Gapless.
Gapless is preferable; but doesn't guarantee long albums/playlists will play to the end.
When Gapless is turned off a slider may be used to adjust the gap in seconds between tracks.
Sufficient time needs to be allowed for each track to load into your player of choice.

The number of seconds needs to be high enough so that tracks are not cut off before they finish playing;
but not so high as to result in unnecessarily long gaps between tracks.

8 seconds seems to work well and is the default. Some experimentation is in order though as each user's systems
and requirements may differ.


plugin.audio.qobuz/changelog.txt
1.1.3
-----
* Fixed bug that was preventing playlists from loading

1.1.2
-----
* Added choice within Audiophile settings between Gapless and non-Gapless
* Gapless doesn't guarantee long albums/playlists will play to the end
* When Gapless is turned off a slider may be used to adjust the gap in seconds between tracks



1.1.1
-----
* Extended Audiophile Option to play individual tracks
* Bugfix to remove invalid import in album.py

1.1.0
-----
* Added Audiophile option - creates playlist opened by external player

1.0.0
-----
* clean up
* bugfixes
* public playlists

0.11.2
-----
* fixes and covers

0.11.0 
-----
* Add "friends" feature
* add http-resolver for musicdb support (experimental)
* add pagination support (for small devices and very big playlists)
* move start/stop api calls in the service
* add "genre" node to browse music
* improve favorite support (add albums & artists & remove)
* lots of cleanup & cosmetics
* use requests 1.0.4
 
0.10.2
-----
* update language files from Transifex

0.10.1
-----
* fix bug in args checking

0.10.0
-----
* big cleaning & rewrites
* add common cache support (disabled)
* speed optimisations

0.9.14
-----
* remove sys.exit since it exit the whole xbmc since frodo rc2


0.9.13
-----
* retrieve and show artists picture (for similar artists)
* disable scan feature by default
* add favorites item (wip)

0.9.12
-----
* fix utf8 glitch

0.9.11
-----
* add context menu item "add as new playlist"
* add option to hide recommendations

0.9.10
-----
* fixes for release
* add favorites support
* remove lastfm similar artists support and use qobuz one

0.9.9
-----
* updates for qobuz api v0.2

0.9.0
-----
* re-add context items (playlist management, track/album/artist navigation, etc.)

release
18/03/2012 12:30 Tag

0.2.0
-----
rewrite

0.1.1
-----
bug fixes

0.1.0
-----
lots of things

0.0.5
-----
* player rewrite (use a backgound process to resolv urls)

0.0.4
-----
* improve caching
* generic tagging
* album view for purchases
* context menu for cache cleaning

0.0.3
-----
* add stop/start for Qobuz accounting
* multiples cleanups & bugfixes
* settings for caches expiration

0.0.2
-----
* [added features] search by artists and albums

0.0.1 
-----
* Where it all began (Support favorite playlists, track search, recommendations)

----- inspired by XBMC Grooveshark Addon (Stephen Denham) -----


plugin.audio.qobuz/resources/lib/qobuz/node/album.py
'''
    qobuz.node.album
    ~~~~~~~~~~~~~~~~

    :copyright: (c) 2012 by Joachim Basmaison, Cyril Leclerc
    :license: GPLv3, see LICENSE for more details.
'''
import os
from inode import INode
from debug import warn
from gui.util import getImage, getSetting, htm2xbmc
from gui.contextmenu import contextMenu
from api import api
from node import getNode, Flag
from sys import platform as _platform
import simplejson as json
import urllib, urllib2
import xbmc
import time
try:
	import win32com.client
except:
	pass
from sys import platform as _platform



SPECIAL_PURCHASES = ['0000020110926', '0000201011300', '0000020120220',
                     '0000020120221']


class Node_album(INode):
    '''
        @class Node_product:
    '''
    def __init__(self, parent, params):
        super(Node_album, self).__init__(parent, params)
        self.nt = Flag.ALBUM
        self.image = getImage('album')
        self.content_type = 'songs'
        self.is_special_purchase = False
        self.offset = None
        self.imageDefaultSize = 'large'
        self.label = 'Album'
        try:
            self.imageDefaultSize = getSetting('image_default_size')
        except:
            pass

        @property
        def nid(self):
            return self._nid
        @nid.getter
        def nid(self):
            return self._nid
        @nid.setter
        def nid(self, value):
            self._id = value
            if value in SPECIAL_PURCHASES:
                self.is_special_purchase = True

    def fetch(self, Dir, lvl, whiteFlag, blackFlag):
        data = api.get('/album/get', album_id=self.nid)
        if not data:
            warn(self, "Cannot fetch product data")
            return False
        self.data = data
        return True

    def populate(self, Dir, lvl, whiteFlag, blackFlag):
    	theUrls = ''
    	launchApp = False
        for track in self.data['tracks']['items']:
            node = getNode(Flag.TRACK)
            if not 'image' in track:
                track['image'] = self.get_image()
            node.data = track
            track_id = track.get('id')
            track_duration = track.get('duration')
            format_id = 6 if getSetting('streamtype') == 'flac' else 5
            data = api.get('/track/getFileUrl', format_id=format_id,
            	track_id=track_id, user_id=api.user_id)
            if not data:
            	theUrls += "Cannot get stream type for track (network problem?)"
            else:
            	if (not 'sample' in (data['url'])):
            		theUrls += str(data['url'])
            		theUrls += '\n'
            		if getSetting('audiophile') == 'true':
    					launchApp = True
            self.add_child(node)
        if launchApp:
            #os.system('TASKKILL /F /IM HQPlayer-desktop.exe')
            qobuzPlaylist = str(os.path.expanduser('~'))
            qobuzPlaylist += '/Music/QobuzNow.m3u8'
            completeName = os.path.abspath(qobuzPlaylist)
            file1 = open(completeName,"w")
            toFile = theUrls
            file1.write(toFile)
            file1.close
            if _platform == "darwin":
        		try:           
        			cmd = """osascript -e 'tell app "HQPlayerDesktop" to quit'"""
        			os.system(cmd)
        			os.system("/Applications/HQPlayerDesktop.app/Contents/MacOS/HQPlayerDesktop "+completeName+"&")
        			cmd = """osascript<<END
        				launch application "System Events"
						tell application "System Events"
							set frontmost of process "HQPlayerDesktop3" to true
						end tell
					END"""
        			os.system(cmd)
        		except:
        			os.system("open "+completeName)
            elif _platform == "win32":
                os.system('TASKKILL /F /IM HQPlayer-desktop.exe')
                os.startfile(completeName, 'open')
        return len(self.data['tracks']['items'])

    def make_url(self, **ka):
        if 'asLocalURL' in ka and ka['asLocalURL']:
            from constants import Mode
            ka['mode'] = Mode.SCAN
        return super(Node_album, self).make_url(**ka)
    
    def makeListItem(self, replaceItems=False):
        import xbmc, xbmcgui
        image = self.get_image()
        thumb = xbmc.getCacheThumbName(image)
        item = xbmcgui.ListItem(
            label=self.get_label(),
            label2=self.get_label(),
            iconImage=image,
            thumbnailImage=image,
            path=self.make_url(),
        )
        item.setInfo('music', infoLabels={
            'genre': self.get_genre(),
            'year': self.get_year(),
            'artist': self.get_artist(),
            'title': self.get_title(),
            'album': self.get_title(),
            'comment': self.get_description()
        })
        ctxMenu = contextMenu()
        self.attach_context_menu(item, ctxMenu)
        item.addContextMenuItems(ctxMenu.getTuples(), replaceItems)
        return item

    '''
    PROPERTIES
    '''
    def get_artist(self):
        return self.get_property(['artist/name',
                               'interpreter/name', 
                               'composer/name'])

    def get_album(self):
        album = self.get_property('name')
        if not album:
            return ''
        return album

    def get_artist_id(self):
        return self.get_property(['artist/id',
                               'interpreter/id',
                              'composer/id'])

    def get_title(self):
        return self.get_property('title')

    def get_image(self, size = None):
        if not size:
            size = self.imageDefaultSize
        return self.get_property(['image/%s' % (size),
                                   'image/large', 
                                   'image/small',
                                   'image/thumbnail'])

    def get_label(self):
        artist = self.get_artist() or 'VA'
        label = '%s - %s' % (artist, self.get_title())
        return label

    def get_genre(self):
        return self.get_property('genre/name')

    def get_year(self):
        import time
        date = self.get_property('released_at')
        year = 0
        try:
            year = time.strftime("%Y", time.localtime(date))
        except:
            pass
        return year

    def get_description(self):
        return htm2xbmc(self.get_property('description'))


plugin.audio.qobuz/resources/lib/qobuz/service/httpd.py
#import string,cgi,time
#from os import curdir, sep
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
from gui.util import lang, getImage, runPlugin, getSetting
import socket
import re
import sys
import urllib2


VERSION = '0.0.1'

username = password = base_path = stream_type = stream_format = None
cache_duration_long = 60*60*24*365
cache_duration_middle = 60*60*24*365
stream_format = 5
__image__ = ''


import xbmcaddon, xbmcplugin, xbmc

import os
pluginId = 'plugin.audio.qobuz'
__addon__ = xbmcaddon.Addon(id=pluginId)
__addonversion__ = __addon__.getAddonInfo('version')
__addonid__ = __addon__.getAddonInfo('id')
__cwd__ = __addon__.getAddonInfo('path')
__addondir__ = __addon__.getAddonInfo('path')
__libdir__ = xbmc.translatePath(os.path.join(__addondir__, 'resources', 'lib'))
__qobuzdir__ = xbmc.translatePath(os.path.join(__libdir__, 'qobuz'))
sys.path.append(__libdir__)
sys.path.append(__qobuzdir__)

from bootstrap import QobuzBootstrap
__handle__ = -1
boot = QobuzBootstrap(__addon__, __handle__)
boot.bootstrap_directories()

def try_get_settings():
    global username, password
    username = __addon__.getSetting('username')
    password = __addon__.getSetting('password')
    if username and password:
        return True
    return False

while not (try_get_settings()):
    xbmc.sleep(5000)

import qobuz
from api import api
from cache import cache
cache.base_path = qobuz.path.cache
api.login(username, password)


stream_format = 6 if __addon__.getSetting('streamtype') == 'flac' else 5
cache_durationm_middle = int(__addon__.getSetting('cache_duration_middle')) * 60
cache_duration_long = int(__addon__.getSetting('cache_duration_long')) * 60

if stream_format == 6:
    stream_mime = 'audio/flac'
else:
    stream_mime = 'audio/mpeg'
    
from debug import log
from node import getNode, Flag


class XbmcAbort(Exception):
    def __init__(self, *a, **ka):
        super(XbmcAbort, self).__init__(*a, **ka)
 
class BadRequest(Exception):
    def __init__(self, *a, **ka):
        self.code = 400
        self.message = 'Bad Request '
        super(BadRequest, self).__init__(*a, **ka)
               
class Unauthorized(Exception):
    def __init__(self, *a, **ka):
        self.code = 401
        self.message = 'Unauthorized '
        super(Unauthorized, self).__init__(*a, **ka)

class RequestFailed(Exception):
    def __init__(self, *a, **ka):
        self.code = 402
        self.message = 'Request Failed'
        super(RequestFailed, self).__init__(*a, **ka)
        
class ServerErrors(Exception):
    def __init__(self, *a, **ka):
        self.code = 500
        self.message = 'Server errors'
        super(ServerErrors, self).__init__(*a, **ka)

class QobuzResponse:
    
    def __init__(self, request):
        self.request = request
        if not self.__parse_path(request.path):
            raise RequestFailed()

    @property
    def path(self):
        return self._path
    @path.getter
    def path(self):
        return self._path
    @path.setter
    def path(self, value):    
        self.reset_request()
        self.__parse_path(value)
        self._path = value

    def reset_request(self):
        self.path = None
        self.request = None
        self.fileExt = None
        self.track_id = None
        self.album_id = None
        self.artist_id = None
        self.fileWanted = None

    def __parse_path(self, path):
        m = re.search('^/qobuz/(\d+)/(.*)$', path)
        if not m:
            return False
        self.artist_id = m.group(1)
        m = re.search('^(\d+)/(.*)$', m.group(2))
        if not m:
            return False
        self.album_id = m.group(1)
        m = re.search('^(\d+|cdart)\.(mp3|mpc|flac|png)', m.group(2))
        if not m:
            return False
        self.track_id = m.group(1)
        self.fileExt = m.group(2)
        self.fileWanted = 'music'
        return True
    
class QobuzHttpResolver_Handler(BaseHTTPRequestHandler):

    def __init__(self, request, client_address, server):
        # FIXME workaround for exceptions in logs when the client broke connexion
        try: 
            BaseHTTPRequestHandler.__init__(self, request, client_address, server)
        except:
            pass
        self.server_version = 'QobuzXbmcHTTP/0.0.1'

    
    def __GET_track(self, request):
        api.login(api.username, api.password)
        node = getNode(Flag.TRACK, {'nid': request.track_id})
        streaming_url = node.get_streaming_url()
        if not streaming_url and getSetting('audiophile') == 'false':
            raise RequestFailed()
        self.send_response(303, "Resolved")
        self.send_header('content-type', stream_mime)
        self.send_header('location', streaming_url)
        self.end_headers()  
    
    def do_GET(self):
        try:
            request = QobuzResponse(self)
            if request.fileWanted == 'music':
                return self.__GET_track(request)
            else:
                raise BadRequest()
        except BadRequest as e: pass
        except Unauthorized as e:
            self.send_error(e.code, e.message)
        except RequestFailed as e:
            self.send_error(e.code, e.message)
        except Exception as e:
            msg = 'Server errors (%s / %s)\n%s' % (
                                                  sys.exc_type, sys.exc_value,
                                                  repr(e))
            self.log_message(msg)
            self.send_error(500, msg)

    def do_HEAD(self):
        try:
            request = QobuzResponse(self)
            if request.fileWanted == 'music':
                self.send_response(200, "Ok ;)")
                self.send_header('content-type', stream_mime)
                self.end_headers()
                return True
            else:
                self.send_response(404, "Nops")
                self.send_header('Content-Type', 'text/html; charset=iso-8859-1')
                self.end_headers()
                return True
        except Unauthorized as e:
            self.send_error(e.code, e.message)
        except RequestFailed as e:
            self.send_error(e.code, e.message)
        except Exception as e:
            msg = 'Server errors (%s / %s)\n%s' % (
                                                  sys.exc_type, sys.exc_value,
                                                  repr(e))
            self.log_message(msg)
            self.send_error(500, msg)

import select
from SocketServer import ThreadingMixIn
import threading

class QobuzHttpResolver(ThreadingMixIn, HTTPServer):

    def abortRequested(self):
        try: 
            return xbmc.abortRequested
        except:
            False
        return False

#    def get_request(self):
#        """Get the request and client address from the socket."""
#        self.socket.settimeout(0.1)
#        result = None
#        while not result:
#            print "TIMEOUT"
#            if self.abortRequested():
#                self.__shutdown_request = True
#                raise KeyboardInterrupt()
#            try:
#                result = self.socket.accept()
#            except socket.timeout:
#                print 'Socket timeout'
#                pass
#        ''' Reset timeout on the new socket '''
#        result[0].settimeout(None)
#        return result

    def verify_request(self, path, client_address): 
        host, port = client_address
        if host == '127.0.0.1': 
            return True
        return False

class QobuzXbmcHttpResolver(QobuzHttpResolver):

    def __init__(self):
        QobuzHttpResolver.__init__(self, ('127.0.0.1', 33574), 
                                                    QobuzHttpResolver_Handler)

def main():
    server = None
    try:
        server = QobuzXbmcHttpResolver()
        log(server, 'Starting Qobuz Resolver')
        server.serve_forever()
    except KeyboardInterrupt:
        print '^C received, shutting down server'
    except XbmcAbort:
        log(server, 'Received xbmc abort... closing')
    finally:
        if server:
            server.socket.close()
            server = None

if __name__ == '__main__':
    main()


plugin.audio.qobuz/resources/lib/qobuz/node/track.py
#     Copyright 2011 Joachim Basmaison, Cyril Leclerc
#
#     This file is part of xbmc-qobuz.
#
#     xbmc-qobuz is free software: you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation, either version 3 of the License, or
#     (at your option) any later version.
#
#     xbmc-qobuz is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with xbmc-qobuz.   If not, see <http://www.gnu.org/licenses/>.
import os
from constants import Mode
from node import Flag, ErrorNoData
from inode import INode
from debug import warn
from gui.util import lang, getImage, runPlugin, getSetting
from gui.contextmenu import contextMenu
from api import api
try:
	import win32com.client
except:
	pass
from sys import platform as _platform

class Node_track(INode):
    '''
        NODE TRACK
    '''
    def __init__(self, parent=None, parameters=None):
        super(Node_track, self).__init__(parent, parameters)
        self.nt = Flag.TRACK
        self.content_type = 'songs'
        self.qobuz_context_type = 'playlist'
        self.is_folder = False
        self.status = None
        self.image = getImage('song')

    def fetch(self, Dir, lvl, whiteFlag, blackFlag):
        if blackFlag & Flag.STOPBUILD == Flag.STOPBUILD:
            return False
        data = api.get('/track/get', track_id=self.nid)
        if not data:
            return False
        self.data = data
        return True
    
    def populate(self, Dir, lvl, whiteFlag, blackFlag):
        Dir.add_node(self)
        return True

    def make_url(self, **ka):
        if 'asLocalURL' in ka and ka['asLocalURL']:
            album_id = self.get_property('album/id')
            if not album_id:
                album_id = self.parent.nid
            url = 'http://127.0.0.1:33574/qobuz/%s/%s/%s.mpc' % (
                    str(self.get_artist_id()),
                    album_id,
                    str(self.nid))
            return url
        if not 'mode' in ka: 
            ka['mode'] = Mode.PLAY 
        return super(Node_track, self).make_url(**ka)

    def get_label(self, sFormat="%a - %t"):
        sFormat = sFormat.replace("%a", self.get_artist())
        sFormat = sFormat.replace("%t", self.get_title())
        sFormat = sFormat.replace("%A", self.get_album())
        sFormat = sFormat.replace("%n", str(self.get_track_number()))
        sFormat = sFormat.replace("%g", self.get_genre())
        return sFormat

    def get_composer(self):
        try:
            return self.get_property('composer/name')
        except:
            return -1

    def get_interpreter(self):
        try:
            return self.get_property('performer/name')
        except:
            return -1

    def get_album(self):
        try:
            album = self.get_property('album/title')
        except:
            return -1
        if album:
            return album
        if not self.parent:
            return ''
        if self.parent.nt & Flag.ALBUM:
            return self.parent.get_title()
        return ''
    
    def get_album_id(self):
        aid = self.get_property('album/id')
        if not aid and self.parent:
            return self.parent.nid
        return aid
    
    def get_image(self):
        image = self.get_property(['album/image/large', 'image/large', 
                                      'image/small',
                                      'image/thumbnail', 'image'])
        if image:
            return image.replace('_230.', '_600.')
        if not self.parent:
            return self.image
        if self.parent.nt & (Flag.ALBUM | Flag.PLAYLIST):
            return self.parent.get_image()

    def get_playlist_track_id(self):
        return self.get_property('playlist_track_id')

    def get_position(self):
        return self.get_property('position')

    def get_title(self):
        return self.get_property('title')

    def get_genre(self):
        genre = self.get_property('album/genre/name')
        if genre:
            return genre
        if not self.parent:
            return ''
        if self.parent.nt & Flag.ALBUM:
            return self.parent.get_genre()
        return ''

    def get_streaming_url(self):
    	theUrls = ''
    	launchApp = False
    	#if getSetting('audiophile') == 'true':
    					#launchApp = True
    	#setSetting(id="audiophile", value='false')
        data = self.__getFileUrl()
        if not data:
            return False
        restrictions = self.get_restrictions()
        for restriction in restrictions:
            print "Restriction: %s" % (restriction)
        if not 'url' in data:
            warn(self, "streaming_url, no url returned\n"  
                "API Error: %s" % (api.error)) 
            return None
        if not data:
            	theUrls += "Cannot get stream type for track (network problem?)"
            	launchApp = False
        else:
            if (not 'sample' in (data['url'])):
            	theUrls = str(data['url'])
            	theUrls += '\n'
            	if getSetting('audiophile') == 'true':
    					launchApp = True
    	if launchApp:
        	qobuzPlaylist = str(os.path.expanduser('~'))
        	qobuzPlaylist += '/Music/QobuzNow.m3u8'
        	completeName = os.path.abspath(qobuzPlaylist)
        	#file1 = open(completeName,"r")
        	#theLines=file1.read()
        	#file1.close
        	#if not (data['url'] in theLines):
        	file1 = open(completeName,"w")
        	file1.write(theUrls)
        	file1.close
        		#setSetting(id="audiophile", value='true')
        	if _platform == "darwin":
        		try:           
        			cmd = """osascript -e 'tell app "HQPlayerDesktop" to quit'"""
        			os.system(cmd)
        			os.system("/Applications/HQPlayerDesktop.app/Contents/MacOS/HQPlayerDesktop "+completeName+"&")
        			cmd = """osascript<<END
        				launch application "System Events"
						tell application "System Events"
							set frontmost of process "HQPlayerDesktop3" to true
						end tell
					END"""
        			os.system(cmd)
        		except:
        			os.system("open "+completeName)
        	elif _platform == "win32":
        		os.system('TASKKILL /F /IM HQPlayer-desktop.exe')
        		os.startfile(completeName, 'open')
        	return False
        else:
        	return data['url']

    def get_artist(self):
        return self.get_property(['artist/name',
                               'composer/name',
                               'performer/name',
                               'interpreter/name',
                               'composer/name',
                               'album/artist/name'])

    def get_artist_id(self):
        s = self.get_property(['artist/id',
                               'composer/id',
                               'performer/id',
                               'interpreter/id'])
        if s:
            return int(s)
        return None

    def get_track_number(self):
        return self.get_property('track_number')

    def get_media_number(self):
        return self.get_property('media_number')

    def get_duration(self):
        duration = self.get_property('duration')
        if duration and int(duration) != 0:
            return duration
        else:
            return -1

    def get_year(self):
        import time
        try:
            date = self.get_property('album/released_at')
            if not date and self.parent and self.parent.nt & Flag.ALBUM:
                return self.parent.get_year()
        except:
            pass
        year = 0
        try:
            year = time.strftime("%Y", time.localtime(date))
        except:
            pass
        return year
    
    def is_playable(self):
        url = self.get_streaming_url()
        if not url:
            return False
        restrictions = self.get_restrictions()
        if 'TrackUnavailable' in restrictions:
            return False
        if 'AlbumUnavailable' in restrictions:
            return False
        return True
    
    def get_description(self):
        if self.parent:
            return self.parent.get_description()
        return ''

    def __getFileUrl(self):
        format_id = 6 if getSetting('streamtype') == 'flac' else 5
        data = api.get('/track/getFileUrl', format_id=format_id,
                           track_id=self.nid, user_id=api.user_id)
        if not data:
            warn(self, "Cannot get stream type for track (network problem?)")
            return None
        return data

    def get_restrictions(self):
        data = self.__getFileUrl()
        if not data: 
            raise ErrorNoData('Cannot get track restrictions')
        restrictions = []
        if not 'restrictions' in data:
            return restrictions
        for restriction in data['restrictions']:
            restrictions.append(restriction['code'])
        return restrictions

    def is_sample(self):
        data = self.__getFileUrl()
        if not data:
            return False
        if 'sample' in data:
            return data['sample']
        return False
    
    def get_mimetype(self):
        data = self.__getFileUrl()
        if not data:
            return False
        if not 'format_id' in data:
            warn(self, "Cannot get mime/type for track (restricted track?)")
            return False
        formatId = int(data['format_id'])
        mime = ''
        if formatId == 6:
            mime = 'audio/flac'
        elif formatId == 5:
            mime = 'audio/mpeg'
        else:
            warn(self, "Unknow format " + str(formatId))
            mime = 'audio/mpeg'
        return mime

    """ We add this information only when playing item because it require
        us to fetch data from Qobuz
    """
    def item_add_playing_property(self, item):
        mime = self.get_mimetype()
        if not mime:
            warn(self, "Cannot set item streaming url")
            return False
        item.setProperty('mimetype', mime)
        item.setPath(self.get_streaming_url())
        return True

    def makeListItem(self, replaceItems=False):
        import xbmcgui
        media_number = self.get_media_number()
        if not media_number:
            media_number = 1
        else:
            media_number = int(media_number)
        duration = self.get_duration()
        if duration == -1:
            import pprint
            print "Error: no duration\n%s" % (pprint.pformat(self.data))
        label = self.get_label()
        isplayable = 'true'

        # Disable free account checking here, purchased track are
        # still playable even with free account, but we don't know yet.
        # if qobuz.gui.is_free_account():
        #    duration = 60
        # label = '[COLOR=FF555555]' + label + '[/COLOR]
        # [[COLOR=55FF0000]Sample[/COLOR]]'
        mode = Mode.PLAY
        url = self.make_url(mode=mode)
        image = self.get_image()
        item = xbmcgui.ListItem(label,
                                label,
                                image,
                                image,
                                url)
        item.setIconImage(image)
        item.setThumbnailImage(image)
        if not item:
            warn(self, "Cannot create xbmc list item")
            return None
        item.setPath(url)
        track_number = self.get_track_number()
        if not track_number:
            track_number = 0
        else:
            track_number = int(track_number)
        mlabel = self.get_property('label/name')
        description = self.get_description()
        comment = ''
        if mlabel:
            comment = mlabel
        if description:
            comment += ' - ' + description
        '''Xbmc Library fix: Compilation showing one entry by track
            We are setting artist like 'VA / Artist'
            Data snippet:
                {u'id': 26887, u'name': u'Interpr\xe8tes Divers'}
                {u'id': 145383, u'name': u'Various Artists'}
                {u'id': 255948, u'name': u'Multi Interpretes'}
        '''
        artist = self.get_artist()
        if self.parent and hasattr(self.parent, 'get_artist_id'):
            artist_id = str(self.parent.get_artist_id())
            #if artist_id in ['26887', '145383', '255948']:
            if self.parent.get_artist() != artist:
                artist = '%s / %s' % (self.parent.get_artist(), artist)
        desc = description or 'Qobuz Music Streaming'
        item.setInfo(type='music', infoLabels={
                     'count': self.nid,
                     'title': self.get_title(),
                     'album': self.get_album(),
                     'genre': self.get_genre(),
                     'artist': artist,
                     'tracknumber': track_number,
                     'duration': duration,
                     'year': self.get_year(),
                     'comment': desc + ' (aid=' + self.get_album_id() + ')'
                     # 'lyrics': "Chant down babylon lalalala" 
                     })
        item.setProperty('DiscNumber', str(media_number))
        item.setProperty('IsPlayable', isplayable)
        item.setProperty('IsInternetStream', isplayable)
        item.setProperty('Music', isplayable)
#        item.setProperty('mimetype', 'audio/mpeg')
        ctxMenu = contextMenu()
        self.attach_context_menu(item, ctxMenu)
        item.addContextMenuItems(ctxMenu.getTuples(), replaceItems)
        return item

    def attach_context_menu(self, item, menu):
        if self.parent and (self.parent.nt & Flag.PLAYLIST == Flag.PLAYLIST):
            colorCaution = getSetting('item_caution_color')
            url = self.parent.make_url(nt=Flag.PLAYLIST,
                id=self.parent.nid,
                qid=self.get_playlist_track_id(),
                nm='gui_remove_track',
                mode=Mode.VIEW)
            menu.add(path='playlist/remove', 
                     label=lang(30073),
                     cmd=runPlugin(url), color=colorCaution)

        ''' Calling base class '''
        super(Node_track, self).attach_context_menu(item, menu)
